File: Game/Network/LeaderboardAPI.swift
import Foundation

@MainActor
final class LeaderboardAPI {
    static let shared = LeaderboardAPI()

    #if DEBUG
    private let baseURL = "https://uzls-leaderboard.a-kuz.workers.dev"
    #else
    private let baseURL = "https://uzls-leaderboard.a-kuz.workers.dev"
    #endif

    private let ud = UserDefaults.standard
    private let session: URLSession = {
        let cfg = URLSessionConfiguration.default
        cfg.timeoutIntervalForRequest = 8
        cfg.timeoutIntervalForResource = 15
        cfg.waitsForConnectivity = false
        return URLSession(configuration: cfg)
    }()

    var playerId: String {
        get {
            if let id = ud.string(forKey: "lb.playerId") { return id }
            let id = UUID().uuidString.lowercased()
            ud.set(id, forKey: "lb.playerId")
            return id
        }
        set { ud.set(newValue, forKey: "lb.playerId") }
    }

    var username: String {
        get { ud.string(forKey: "lb.username") ?? "" }
        set { ud.set(newValue, forKey: "lb.username") }
    }

    private var flushing = false

    private init() {}

    // MARK: - Pending queue (persisted)

    struct PendingResult: Codable {
        let levelId: Int
        let moves: Int
        let timeMs: Int
        let date: Date
    }

    private var pendingQueue: [PendingResult] {
        get {
            guard let data = ud.data(forKey: "lb.pending") else { return [] }
            return (try? JSONDecoder().decode([PendingResult].self, from: data)) ?? []
        }
        set {
            ud.set(try? JSONEncoder().encode(newValue), forKey: "lb.pending")
        }
    }

    private func enqueue(levelId: Int, moves: Int, timeMs: Int) {
        var q = pendingQueue
        q.append(PendingResult(levelId: levelId, moves: moves, timeMs: timeMs, date: Date()))
        pendingQueue = q
    }

    // MARK: - Player (stateless — just generates username if needed)

    func ensureRegistered() async {
        if !username.isEmpty {
            await flushQueue()
            return
        }
        do {
            let resp: PlayerResponse = try await post("/api/player", body: ["id": playerId])
            username = resp.username
            await flushQueue()
        } catch {
            print("[Leaderboard] registration failed: \(error)")
        }
    }

    struct PlayerResponse: Decodable {
        let id: String
        let username: String
    }

    func loginAs(_ name: String) {
        let newId = UUID().uuidString.lowercased()
        playerId = newId
        username = name
    }

    func setUsername(_ newName: String) {
        username = newName
    }

    // MARK: - Submit

    struct SubmitResponse: Decodable {
        let total_completions: Int
        let avg_moves: Double
        let best_moves: Int
        let player_best: Int?
        let percentile: Int?
        let is_personal_best: Bool
    }

    @discardableResult
    func submitResult(levelId: Int, moves: Int, timeMs: Int) async -> SubmitResponse? {
        let body: [String: Any] = [
            "player_id": playerId,
            "username": username.isEmpty ? "anon" : username,
            "level_id": levelId,
            "moves": moves,
            "time_ms": timeMs,
        ]

        do {
            let resp: SubmitResponse = try await post("/api/submit", body: body)
            return resp
        } catch {
            print("[Leaderboard] submit failed, queued: \(error)")
            enqueue(levelId: levelId, moves: moves, timeMs: timeMs)
            return nil
        }
    }

    // MARK: - Flush queue

    func flushQueue() async {
        guard !flushing else { return }
        let queue = pendingQueue
        guard !queue.isEmpty else { return }
        flushing = true
        defer { flushing = false }

        var remaining: [PendingResult] = []
        for item in queue {
            let body: [String: Any] = [
                "player_id": playerId,
                "username": username.isEmpty ? "anon" : username,
                "level_id": item.levelId,
                "moves": item.moves,
                "time_ms": item.timeMs,
            ]
            do {
                let _: SubmitResponse = try await post("/api/submit", body: body)
            } catch {
                remaining.append(item)
            }
        }
        pendingQueue = remaining
    }

    // MARK: - Stats

    struct StatsResponse: Decodable {
        let total_completions: Int
        let avg_moves: Double
        let best_moves: Int
        let player_best: Int?
        let percentile: Int?
    }

    func fetchStats(levelId: Int, moves: Int? = nil) async -> StatsResponse? {
        var url = "\(baseURL)/api/stats/\(levelId)?player_id=\(playerId)"
        if let m = moves { url += "&moves=\(m)" }
        return try? await get(url)
    }

    // MARK: - Networking

    enum LeaderboardError: Error {
        case httpError(Int)
    }

    private func post<T: Decodable>(_ path: String, body: Any) async throws -> T {
        let url = URL(string: "\(baseURL)\(path)")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) else {
            let code = (response as? HTTPURLResponse)?.statusCode ?? 0
            throw LeaderboardError.httpError(code)
        }
        return try JSONDecoder().decode(T.self, from: data)
    }

    private func get<T: Decodable>(_ urlString: String) async throws -> T {
        let url = URL(string: urlString)!
        let request = URLRequest(url: url)

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) else {
            let code = (response as? HTTPURLResponse)?.statusCode ?? 0
            throw LeaderboardError.httpError(code)
        }
        return try JSONDecoder().decode(T.self, from: data)
    }
}

File: android/app/src/main/java/com/uzls/four/ui/LeaderboardAPI.kt
package com.uzls.four.ui

import android.content.Context
import android.content.SharedPreferences
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.net.HttpURLConnection
import java.net.URL
import java.util.UUID

class LeaderboardAPI(context: Context) {
    companion object {
        private const val BASE_URL = "https://uzls-leaderboard.a-kuz.workers.dev"
    }

    private val prefs: SharedPreferences =
        context.getSharedPreferences("uzls_leaderboard", Context.MODE_PRIVATE)
    private val json = Json { ignoreUnknownKeys = true; isLenient = true }

    var playerId: String
        get() {
            val id = prefs.getString("playerId", null)
            if (id != null) return id
            val newId = UUID.randomUUID().toString()
            prefs.edit().putString("playerId", newId).apply()
            return newId
        }
        set(value) { prefs.edit().putString("playerId", value).apply() }

    var username: String
        get() = prefs.getString("username", "") ?: ""
        set(value) { prefs.edit().putString("username", value).apply() }

    @Serializable
    data class PlayerResponse(val id: String, val username: String)

    @Serializable
    data class SubmitResponse(
        val total_completions: Int = 0,
        val avg_moves: Double = 0.0,
        val best_moves: Int = 0,
        val player_best: Int? = null,
        val percentile: Int? = null,
        val is_personal_best: Boolean = false,
    )

    @Serializable
    data class StatsResponse(
        val total_completions: Int = 0,
        val avg_moves: Double = 0.0,
        val best_moves: Int = 0,
        val player_best: Int? = null,
        val percentile: Int? = null,
    )

    suspend fun ensureRegistered() {
        if (username.isNotEmpty()) return
        try {
            val body = """{"id":"$playerId"}"""
            val resp = post("/api/player", body)
            val player = json.decodeFromString<PlayerResponse>(resp)
            username = player.username
        } catch (_: Exception) { }
    }

    suspend fun submitResult(levelId: Int, moves: Int, timeMs: Int): SubmitResponse? {
        val name = username.ifEmpty { "anon" }
        val body = """{"player_id":"$playerId","username":"$name","level_id":$levelId,"moves":$moves,"time_ms":$timeMs}"""
        return try {
            val resp = post("/api/submit", body)
            json.decodeFromString<SubmitResponse>(resp)
        } catch (_: Exception) {
            enqueue(levelId, moves, timeMs)
            null
        }
    }

    suspend fun fetchStats(levelId: Int, moves: Int): StatsResponse? {
        return try {
            val resp = get("/api/stats/$levelId?player_id=$playerId&moves=$moves")
            json.decodeFromString<StatsResponse>(resp)
        } catch (_: Exception) { null }
    }

    fun loginAs(name: String) {
        playerId = UUID.randomUUID().toString()
        username = name
    }

    private fun enqueue(levelId: Int, moves: Int, timeMs: Int) {
        val pending = prefs.getString("pending", "") ?: ""
        val entry = "$levelId,$moves,$timeMs"
        prefs.edit().putString("pending", if (pending.isEmpty()) entry else "$pending|$entry").apply()
    }

    suspend fun flushQueue() {
        val pending = prefs.getString("pending", "") ?: ""
        if (pending.isEmpty()) return
        val entries = pending.split("|")
        val remaining = mutableListOf<String>()
        for (entry in entries) {
            val parts = entry.split(",")
            if (parts.size != 3) continue
            try {
                val name = username.ifEmpty { "anon" }
                val body = """{"player_id":"$playerId","username":"$name","level_id":${parts[0]},"moves":${parts[1]},"time_ms":${parts[2]}}"""
                post("/api/submit", body)
            } catch (_: Exception) {
                remaining.add(entry)
            }
        }
        prefs.edit().putString("pending", remaining.joinToString("|")).apply()
    }

    private suspend fun post(path: String, body: String): String = withContext(Dispatchers.IO) {
        val conn = URL("$BASE_URL$path").openConnection() as HttpURLConnection
        conn.requestMethod = "POST"
        conn.setRequestProperty("Content-Type", "application/json")
        conn.connectTimeout = 8000
        conn.readTimeout = 8000
        conn.doOutput = true
        conn.outputStream.use { it.write(body.toByteArray()) }
        conn.inputStream.bufferedReader().readText()
    }

    private suspend fun get(path: String): String = withContext(Dispatchers.IO) {
        val conn = URL("$BASE_URL$path").openConnection() as HttpURLConnection
        conn.connectTimeout = 8000
        conn.readTimeout = 8000
        conn.inputStream.bufferedReader().readText()
    }
}
